src/GraphQL/Resolver.hs

```diff
@@ -326,3 +336,123 @@ instance forall typeName interfaces fields m.
+
+
+-- | The @Handler@ type of a @API.Union@ must be the same for all
+-- possible Objects, but each Object has a different type. We
+-- unsafeCoerce the return type into an Any, tagging it with the union
+-- and the underlying monad for type safety, but we elide the Object
+-- type itself. This way we can represent all @Handler@ types of the
+-- Union with a single type and still stay type-safe.
+type role DynamicUnionValue representational representational
+data DynamicUnionValue (union :: Type) (m :: Type -> Type) = DynamicUnionValue { _label :: Text, _value :: GHC.Exts.Any }
+
+class RunUnion m union objects where
+  runUnion :: DynamicUnionValue union m -> Selection -> m (Maybe (Result GValue.Value)) -- TODO return Result
+
+instance forall m union os n i f.
+  ( Monad m
+  , KnownSymbol n
+  , TypeIndex m (API.Object n i f) union ~ Handler m (API.Object n i f)
+  , RunFields m (RunFieldsType m f)
+  , RunUnion m union os
+  ) => RunUnion m union ((API.Object n i f):os) where
+  runUnion duv fragment@(SelectionInlineFragmentPattern _ selection) =
+    case extractUnionValue @(API.Object n i f) @union @m duv of
+      Just handler -> Just <$> (buildResolver @m @(API.Object n i f) handler selection)
+      Nothing -> runUnion @m @union @os duv fragment
+  runUnion _ _ = pure Nothing
+
+-- TODO(tom): AFAICT it should not be possible to ever hit the empty
+-- case because the compiler doesn't allow constructing a unionValue
+-- that's not in the Union. If the following code ever gets executed
+-- it's almost certainly a bug in the union code.
+--
+-- We still need to implement this instance for the compiler because
+-- it exhaustively checks all cases when deconstructs the Union.
+instance forall m union. RunUnion m union '[] where
+  runUnion _ _ = notImplemented
+
+instance forall m unionName objects.
+  ( Monad m
+  , KnownSymbol unionName
+  , RunUnion m (API.Union unionName objects) objects
+  ) => HasGraph m (API.Union unionName objects) where
+  type Handler m (API.Union unionName objects) = m (DynamicUnionValue (API.Union unionName objects) m)
+  buildResolver mHandler selectionSet = do
+    duv@(DynamicUnionValue _label _) <- mHandler
+    -- we only need to look at the fragment that matches by name:
+    case find (matchFragmentName _label) selectionSet of
+      Nothing -> pure (Result [UnionTypeNotFound] GValue.ValueNull) -- TODO more error detail
+      Just fragment -> do
+        -- loop through union handlers and call right one when type matches.
+        r <- runUnion @m @(API.Union unionName objects) @objects duv fragment
+        case r of
+          Just r' -> pure r'
+          Nothing -> pure (Result [ResolveNonUnionType] GValue.ValueNull)
+
+      where
+        matchFragmentName label' (SelectionInlineFragmentPattern (AST.NamedType name') _) =
+          label' == AST.getNameText name'
+        matchFragmentName _ _ = False
+
+
+symbolText :: forall ks. KnownSymbol ks => Text
+symbolText = toS (symbolVal @ks Proxy)
```

Comment:
It's not the same, the other one is

```
nameFromSymbol :: forall (n :: Symbol). KnownSymbol n => Either NameError Name
nameFromSymbol = makeName (toS (symbolVal @n Proxy))
```

I could write `nameFromSymbol` in terms of `symbolText` but not sure it's worth it